### PR TITLE
Improve `Tree` performance

### PR DESCRIPTION
**Pull request by aaronp64**
Added `TreeItem::last_child` to avoid needing to iterate through all children to get to the end.  This mainly helps in cases where one `TreeItem` has many children (1000s), and new children are added to the end, as each add had to iterate through all previously added children.

Gdscript example to compare creating `TreeItem`s, time reduced from 1640ms to 66ms:

```gdscript
func _ready() -> void:
	var root = $Tree.create_item()
	root.set_text(0, "Root")
	var start_time = Time.get_ticks_msec()
	for i in 20000:
		var item = $Tree.create_item()
		item.set_text(0, "Item " + str(i))
	var end_time = Time.get_ticks_msec()
	print("Time: %dms" % (end_time - start_time))
```

This also impacts `SceneTreeEditor`.  I created a `Node2D` scene with 10,000 empty `Node2D` children to test with.  Opening this scene took 4.28 seconds before this change, 1.55 seconds after.  At 20,000 children, time was reduced from 16 seconds to 2.65 seconds.

`EditorDebuggerTree` has similar slowness in adding items, but also is repeatedly updated when running game from the editor.  Using the gdscript below to create 30,000 `Node2D`s at runtime, the editor is significantly more responsive when using the remote scene tree while the parent node stays collapsed.  Once the parent node is expanded, more time is spent handling the tree's text/heights, which gets pretty slow, but is still faster than before.

```gdscript
func _ready() -> void:
	for i in 30000:
		add_child(Node2D.new())
```

Also added some unit tests.  All tests pass with both old and new code, except for the `Clear items` test, which failed on old code due to `children_cache` not being cleared in `clear_children()`.  Looks like we avoided needing to use the cache in existing code where `clear_children()` is called, so this hasn't been an issue.

### CODE REVIEW

**Review by Calinou — APPROVED** [review]
Tested locally, it works as expected.

**Testing project:** [test-tree.zip](https://github.com/user-attachments/files/16385321/test-tree.zip)

## Benchmark

Using an optimized editor build (MSVC 2022 `optimize=speed`). Running the MRP above which contains the first code sample from OP.

Before | After
-|-
745 ms | 38 ms

<details open>

<summary>PC specifications</summary>

- **CPU:** Intel Core i9-13900K
- **GPU:** NVIDIA GeForce RTX 4090
- **RAM:** 64 GB (2×32 GB DDR5-5800 C30)
- **SSD:** Solidigm P44 Pro 2 TB
- **OS:** Windows 11 23H2
</details>

File: tests/test_main.cpp [hunk 1]
```diff
@@ -121,6 +121,7 @@
 #include "tests/scene/test_sprite_frames.h"
 #include "tests/scene/test_theme.h"
 #include "tests/scene/test_timer.h"
+#include "tests/scene/test_tree.h"
```
Comment (AThousandShips) [inline tests/test_main.cpp:124]: Should be moved down to the `#ifndef ADVANCED_GUI_DISABLED` part as it is advanced UI

**Comment by Mickeon** [comment]
It can probably be cherrypicked later, assuming lack of unforeseen consequences

**Comment by Saul2022** [comment]
Wouldn't this fix https://github.com/godotengine/godot/issues/94648#issue-2424447273  along with rthe closing pr?

**Comment by aaronp64** [comment]
> Wouldn't this fix [#94648 (comment)](https://github.com/godotengine/godot/issues/94648#issue-2424447273) along with rthe closing pr?

No, I ran into this when looking into that issue, but most of the time spent in that case is in `DynamicBVH` functions.  From their screenshots, I think the reason `Tree` wasn't an issue there is that the block nodes were internal to the map, so they weren't being added to the `SceneTreeEditor`

File: tests/test_main.cpp (hunk 1 above)
Comment (aaronp64) [reply tests/test_main.cpp]: Moved to advanced gui section

**Review by KoBeWi — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by AThousandShips** [comment]
This broke trees in at least `EditorHelpSearch`, will try to identify what was broken

**Pull request merged**